les/utils/limiter.go

```diff
@@ -0,0 +1,395 @@
+func flatWeight(item interface{}) uint64 { return item.(*nodeQueue).flatWeight }
+
+// add adds the node queue to the address group. It is the caller's responsibility to
+// add the address group to the address map and the address selector if it wasn't
+// there before.
+func (ag *addressGroup) add(nq *nodeQueue) {
+	if nq.groupIndex != -1 {
+		panic("added node queue is already in an address group")
+	}
+	l := len(ag.nodes)
+	nq.groupIndex = l
+	ag.nodes = append(ag.nodes, nq)
+	ag.sumFlatWeight += nq.flatWeight
+	ag.groupWeight = ag.sumFlatWeight / uint64(l+1)
+	ag.nodeSelect.Update(ag.nodes[l])
+}
+
+// update updates the selection weight of the node queue inside the address group.
+// It is the caller's responsibility to update the group's selection weight in the
+// address selector.
+func (ag *addressGroup) update(nq *nodeQueue, weight uint64) {
+	if nq.groupIndex == -1 || nq.groupIndex >= len(ag.nodes) || ag.nodes[nq.groupIndex] != nq {
+		panic("updated node queue is not in this address group")
+	}
+	ag.sumFlatWeight += weight - nq.flatWeight
+	nq.flatWeight = weight
+	ag.groupWeight = ag.sumFlatWeight / uint64(len(ag.nodes))
+	ag.nodeSelect.Update(nq)
+}
+
+// remove removes the node queue from the address group. It is the caller's responsibility
+// to remove the address group from the address map if it is empty.
+func (ag *addressGroup) remove(nq *nodeQueue) {
+	if nq.groupIndex == -1 || nq.groupIndex >= len(ag.nodes) || ag.nodes[nq.groupIndex] != nq {
+		panic("removed node queue is not in this address group")
+	}
+
+	l := len(ag.nodes) - 1
+	if nq.groupIndex != l {
+		ag.nodes[nq.groupIndex] = ag.nodes[l]
+		ag.nodes[nq.groupIndex].groupIndex = nq.groupIndex
+	}
+	nq.groupIndex = -1
+	ag.nodes = ag.nodes[:l]
+	ag.sumFlatWeight -= nq.flatWeight
+	if l >= 1 {
+		ag.groupWeight = ag.sumFlatWeight / uint64(l)
+	} else {
+		ag.groupWeight = 0
+	}
+	ag.nodeSelect.Remove(nq)
+}
+
+// choose selects one of the node queues belonging to the address group
+func (ag *addressGroup) choose() *nodeQueue {
+	return ag.nodeSelect.Choose().(*nodeQueue)
+}
+
+// NewLimiter creates a new Limiter
+func NewLimiter(sumWeightLimit uint) *Limiter {
+	l := &Limiter{
+		addressSelect:  NewWeightedRandomSelect(func(item interface{}) uint64 { return item.(*addressGroup).groupWeight }),
+		valueSelect:    NewWeightedRandomSelect(func(item interface{}) uint64 { return item.(*nodeQueue).valueWeight }),
+		nodes:          make(map[enode.ID]*nodeQueue),
+		addresses:      make(map[string]*addressGroup),
+		sumWeightLimit: sumWeightLimit,
+	}
+	l.cond = sync.NewCond(&l.lock)
+	go l.processLoop()
+	return l
+}
+
+// selectionWeights calculates the selection weights of a node for both the address and
+// the value selector. The selection weight depends on the next request weight or the
+// summed weights of recently dropped requests. relWeight is reqWeight/maxWeight.
+func selectionWeights(relWeight, value float64) (flatWeight, valueWeight uint64) {
```

Comment:
About selection weights: I realized that the terminology is a bit misleading because two things are called "weight". I'll think about better names to clean up the confusion. The weight of a request basically means a fixed cost. Note that we don't use CPU time as a cost metric any more, it can be a constant value depending on the request type (or a sum of fixed costs if requests are grouped in a batch).
The other "weight" is the selection weight which is is used by WeightedRandomSelect and it represents the selection probability. The two are actually inversely related, if there are two requests queued and one of them is considered 10 times heavier then it is selected with 10 times less probability. The `flatWeight` is `100000000/(requestWeight/sumRequestWeights)`. This is good because if there are many "light" and "heavy" requests then all will be served eventually but the heavy requests won't starve the light ones by slowing down the processing queue. If you send a heavy request in a crowded situation, you should be prepared that probably 10 light requests will be served before yours gets selected.
The `valueWeight` is also proportional to `value` which is an arbitrary number assigned to the sending node (maybe the name is also not the most intuitive here). It is used to ensure that once a node has been proven useful it can get a good chance even if there is a DDoS attack and there requests coming from many addresses. You can think of these as "trusted" nodes. Let's imagine a situation where request weights are all the same, there are 10000 requesting nodes and 5 of them are trusted. The address selector selects any of the requesting nodes with 1/10000 chance while the value selector selects one of the trusted nodes with 1/5 chance (assuming that they have the same "value", the same level of trust). The two selectors are both used with 1/2 chance so the good nodes still get 1/10 chance while the attackers get 1/20000. This is good because a DDoS attack can at least not starve the proven useful connections.
In reality the "trustedness" of a node is not necessarily a true/false thing so instead a 0..1 "value" is used and a node can get into the value selector with a proportional chance. It also makes sense to take the individual request weights into account similarly to `flatWeight`. Therefore `valueWeight` is calculated as `100000000*nodeValue/(requestWeight/sumRequestWeights)`. Note that `flatWeight` and `valueWeight` are never compared to each other because they are used in different selectors.

Reply:
Thank you! Now I understand much better!